src/meta/src/serving/mod.rs

```diff
@@ -158,13 +171,43 @@ pub async fn start_serving_vnode_mapping_worker(
     notification_manager: NotificationManagerRef,
     metadata_manager: MetadataManager,
     serving_vnode_mapping: ServingVnodeMappingRef,
+    session_params: SessionParamsControllerRef,
 ) -> (JoinHandle<()>, Sender<()>) {
     let (local_notification_tx, mut local_notification_rx) = tokio::sync::mpsc::unbounded_channel();
     let (shutdown_tx, mut shutdown_rx) = tokio::sync::oneshot::channel();
     notification_manager
         .insert_local_sender(local_notification_tx)
         .await;
     let join_handle = tokio::spawn(async move {
+        let reset = || async {
+            let (workers, streaming_parallelisms) = fetch_serving_infos(&metadata_manager).await;
+            let max_serving_parallelism = session_params
+                .get_params()
+                .await
+                .batch_parallelism()
+                .map(|p| p.get());
+            let (mappings, failed) = serving_vnode_mapping.upsert(
+                streaming_parallelisms,
+                &workers,
+                max_serving_parallelism,
+            );
+            tracing::debug!(
+                "Update serving vnode mapping snapshot for fragments {:?}.",
+                mappings.keys()
+            );
+            if !failed.is_empty() {
+                tracing::warn!(
+                    "Fail to update serving vnode mapping for fragments {:?}.",
+                    failed
+                );
+            }
```

Comment:
Under what circumstances would the failed branch be hit?

Reply:
Only on some bug.
The branch is added because it calls the pure function `place_vnode` can fail, e.g. due to incorrect arguments.